### PR TITLE
mtls support: Login with certificate

### PR DESCRIPTION
**Pull request by EpsilonTal**
Support mTLS login on smctl using a client-id, certificate, and a private key.

### CODE REVIEW

File: internal/cmd/login/login.go [hunk 1]
```diff
@@ -81,6 +83,8 @@ func (lc *Cmd) Prepare(prepare cmd.PrepareFunc) *cobra.Command {
 	result.Flags().StringVarP(&lc.password, "password", "p", "", "Password")
 	result.Flags().StringVarP(&lc.clientID, "client-id", "", "", "Client id used for OAuth flow")
 	result.Flags().StringVarP(&lc.clientSecret, "client-secret", "", defaultClientSecret, "Client secret used for OAuth flow")
+	result.Flags().StringVarP(&lc.cert, "cert", "", "", "Client certificate")
+	result.Flags().StringVarP(&lc.key, "key", "", "", "Client private key")
```
Comment (EpsilonTal) [inline internal/cmd/login/login.go:87]: ua review

File: internal/util/error.go [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ *    limitations under the License.
+ */
+
+package util
+
+import "errors"
+
+var (
+	LoginValidationError     = errors.New("when using client credentials flow, required either client_id & client_secret or client_id & cert & key")
+	MTLSKeyPairCreationError = errors.New("error creating x509 keypair from client cert file and client key file")
```
Comment (EpsilonTal) [inline internal/util/error.go:23]: ua review

File: pkg/auth/oidc/oidc.go [hunk 3]
```diff
@@ -44,19 +49,37 @@ type OpenIDStrategy struct {
 
 // NewOpenIDStrategy returns OpenId auth strategy
 func NewOpenIDStrategy(options *auth.Options) (*OpenIDStrategy, *auth.Options, error) {
-	httpClient := util.BuildHTTPClient(options.SSLDisabled)
+	var httpClient *http.Client
+	var err error
+	if len(options.Cert) > 0 && len(options.Key) > 0 {
+		if httpClient, err = util.BuildHTTPClientWithCert(options.Cert, options.Key); err != nil {
```
Comment (pavelmaliy) [inline pkg/auth/oidc/oidc.go:55]: you're checking it twice in the func,
store it in variable -> mtlsEnabled := len(options.Cert) > 0 && len(options.Key) > 0

File: internal/util/error.go (hunk 2 above)
Comment (EpsilonTal) [reply internal/util/error.go]: `MTLSKeyPairCreationError` - removed, returning the original error
`LoginValidationError` - moved to the package as `validationError`

File: internal/util/util.go [hunk 4]
```diff
@@ -67,3 +67,35 @@ func BuildHTTPClient(sslDisabled bool) *http.Client {
 
 	return client
 }
+
+// BuildHTTPClientWithCert builds https mTLS client
+func BuildHTTPClientWithCert(certPath, keyPath string) (*http.Client, error) {
+	client := getClient()
+	cert, err := tls.LoadX509KeyPair(certPath, keyPath)
+	if err != nil {
+		return nil, MTLSKeyPairCreationError
```
[the remark this answers is not in the document]
Comment (EpsilonTal) [reply internal/util/util.go]: returning the original error now

File: internal/cmd/login/login.go [hunk 5]
```diff
@@ -155,6 +161,9 @@ func (lc *Cmd) Run() error {
 		TokenEndpoint:         options.TokenEndpoint,
 		TokenBasicAuth:        info.TokenBasicAuth,
 	}
+	if len(options.Cert) > 0 && len(options.Key) > 0 {
+		settings.ClientID = options.ClientID
+	}
 	if options.ClientID == defaultClientID && options.ClientSecret == defaultClientSecret {
```
[the remark this answers is not in the document]
Comment (EpsilonTal) [reply internal/cmd/login/login.go]: removed

File: pkg/auth/oidc/oidc.go (hunk 3 above)
Comment (EpsilonTal) [reply pkg/auth/oidc/oidc.go:57]: done

**Review by pavelmaliy — APPROVED** [review]

**Pull request merged**